### PR TITLE
Update u2020 to u2021

### PR DESCRIPTION
**Pull request by cgruber**

### CODE REVIEW

**Comment by felipecsl** [comment]
I'm using this PR as a pre-release version of the upgrade guide for Dagger 2.
Any reason why this PR hasn't been merged? Still work in progress?

**Comment by JakeWharton** [comment]
Dagger is pre-release software.

**Comment by cgruber** [comment]
Hmm.  We should probably maybe merge this in a Dagger2 branch, so it's
available, even if not in main.  Or fork u2020-dagger1.  Something.

**Comment by felipecsl** [comment]
+1 for fork into google/u2021

**Comment by JakeWharton** [comment]
I've always just pointed people who ask at this PR and told them once there
was a release the project would be updated.

**Comment by cgruber** [comment]
Yeah - I think it's just harder to check out and play with it while it's a
PR.  Head of an unreleased branch seems more accessible.  That said, people
CAN just check out my branch.

**Comment by cgruber** [comment]
Side note, we're expecting a release of 2.0 immanently - as in, within a week or less.

File: src/debug/java/com/jakewharton/u2020/U2020Component.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+@Singleton
+@Component(modules = { U2020AppModule.class, DebugUiModule.class, DebugDataModule.class })
+public interface U2020Component extends U2020Graph {
+  /**
+   * An initializer that creates the graph from an application. 
+   */
+  final static class Initializer {
+    static U2020Graph init(U2020App app) {
+      return Dagger_U2020Component.builder()
```
Comment (fourlastor) [inline src/debug/java/com/jakewharton/u2020/U2020Component.java]: this should be updated with the new naming convention `DaggerU2020Component`

File: build.gradle [hunk 2]
```diff
@@ -18,23 +19,27 @@ def versionBuild = 0 // bump for dogfood builds, public betas, etc.
...
 }
 
 dependencies {
   compile 'com.android.support:support-v4:21.0.0'
 
-  compile 'com.squareup.dagger:dagger:1.2.1'
-  provided 'com.squareup.dagger:dagger-compiler:1.2.1'
+  compile 'com.google.dagger:dagger:2.0-SNAPSHOT'
+  apt 'com.google.dagger:dagger-compiler:2.0-SNAPSHOT'
+  provided 'org.glassfish:javax.annotation:10.0-b28'
```
Comment (marcin-kozinski) [inline build.gradle]: Wouldn't it be better/cleaner/more elegant to use `'javax.annotation:javax.annotation-api:1.2'`? The `org.glassfish` group and the weird version number don't feel good, but maybe it's just me.

**Comment by machinezhou** [comment]
@JakeWharton @cgruber  hi, since the dagger2 was released and when does the u2020 version comes out?

**Comment by johntowell** [comment]
yes a u2021 sample app with dagger2 seems useful. Is there one anywhere currently?

**Comment by Jogan** [comment]
@johntowell @machinezhou A did a first pass at implementing the current app with Dagger 2.  It can be found here https://github.com/prolificinteractive/u2020.

**Comment by Kisty** [comment]
Now that Dagger 2 is officially released (and has been for a while). Can you update this?

In the meantime, should I be looking at @prolificinteractive's version or @cgruber's?

**Comment by vaughandroid** [comment]
Looking at this, the approach of overriding `getSystemService()` seems like it's going to break down with Dagger 2. Either you can only have one scope, or you're going to have to do some nasty casting for each `@Component` type. I guess you could write an `Injector` class for each scope, but that starts to feel like boilerplate.

Anyone have any ideas on solving this?

**Comment by vaughandroid** [comment]
I should say, I realise u2020 only has a single scope, but a given the aims of the project it probably shouldn't demonstrate a technique which is going to fall apart as soon as it's used in non-trivial cases.

**Comment by Zhuinden** [comment]
Dagger1 is deprecated https://github.com/square/dagger

**Comment by piotrek1543** [comment]
@JakeWharton @cgruber are you planning to continue this branch? It seems pretty outdated.

**Comment by Zhuinden** [comment]
I doubt it.

**Comment by JakeWharton** [comment]
Closing since it's wildly outdated. Will look to port our Dagger 2 implementation out into this project.

File: build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (solenwin2) [reply build.gradle]: Readme_edit.

File: src/debug/java/com/jakewharton/u2020/U2020Component.java (hunk 1 above)
Comment (solenwin2) [reply src/debug/java/com/jakewharton/u2020/U2020Component.java]: Readme.

**Pull request closed without merging**